src/Annotation.cpp

```diff
@@ -0,0 +1,500 @@
+#include "Annotation.h"
+#include "Debug.h"
+#include "ErrorAnnotation.h"
+#include "FunctionAnnotation.h"
+#include "ParameterAnnotation.h"
+#include "ReturnAnnotation.h"
+#include "StructAnnotation.h"
+#include "Utils.h"
+
+// TODO: error reporting should be more robust and not just a debug logout
+// message; it should potentially be managed by a class that contains the
+// different types of errors
+
+/*
+How correct formatting is defined:
+
+For some string s and some char of string s[i], it can either be alphanumeric
+(alnum), comma, or space, or we could be at the end of the string (EOS)
+
+if s[i] is alphanumeric:
+  -s[i+1] can be alnum, comma, EOS
+  -it can't be space
+if s[i] is comma:
+  -s[i+1] can be space, alnum
+  -it can't be EOS, comma
+if s[i] is space:
+  -s[i+1] can be alnum
+  -it can't be space, EOS, comma
+
+Also, the string cannot be empty and s[0] must be alnum
+
+Example valid strings: "x", "x, y", "x,y,z", "x, y,z", "x, y, z"
+Example invalid strings: " x", "x y", "x,  y", "x,y,z  "
+*/
+bool methodsArgumentIsCorrectlyFormatted(const std::string &rawMethodsString) {
+  int rawMethodsStringSize = rawMethodsString.size();
+
+  if (rawMethodsStringSize == 0) {
+    logout("Method Args Error 0 on '"
+           << rawMethodsString << "': string must not be empty") return false;
+  }
+
+  if (!std::isalnum(rawMethodsString[0])) {
+    logout("Method Args Error 1 on '"
+           << rawMethodsString
+           << "': first character of string must be alphanumeric") return false;
+  }
+
+  for (int i = 0; i < rawMethodsStringSize; i++) {
+    if (rawMethodsString[i] == ',') {
+      if ((i == rawMethodsStringSize - 1) || rawMethodsString[i + 1] == ',') {
+        logout("Method Args Error 2 on '"
+               << rawMethodsString
+               << "': comma must not be last character and a comma cannot "
+                  "follow a comma") return false;
+      }
+    } else if (rawMethodsString[i] == ' ') {
+      if ((i == rawMethodsStringSize - 1) || rawMethodsString[i + 1] == ' ' ||
+          rawMethodsString[i + 1] == ',') {
+        logout("Method Args Error 3 on '"
+               << rawMethodsString
+               << "': space must be followed by alnum and cannot be last char, "
+                  "another space, or a comma") return false;
+      }
+    } else if (std::isalnum(rawMethodsString[i])) {
+      if ((i != rawMethodsStringSize - 1) && rawMethodsString[i + 1] == ' ') {
+        logout(
+            "Method Args Error 4 on '"
+            << rawMethodsString
+            << "': alnum character cannot be followed by space") return false;
+      }
+    }
+  }
+  return true;
+}
+
+bool rawStringIsCorrectlyFormatted(const std::string &rawAnnotationString) {
```

Comment:
We should add a TODO to automate the tests you've written for this method in `test19.c`. I'm guessing you checked that test's output by hand to test this?